src/Orleans.Core.Abstractions/Lifecycle/IGrainLifecycle.cs

```diff
@@ -10,5 +16,78 @@ namespace Orleans.Runtime
     /// </remarks>
     public interface IGrainLifecycle : ILifecycleObservable
     {
+        /// <summary>
+        /// Registers a grain migration participant.
+        /// </summary>
+        /// <param name="participant">The participant.</param>
+        void AddMigrationParticipant(IGrainMigrationParticipant participant);
+
+        /// <summary>
+        /// Unregisters a grain migration participant.
+        /// </summary>
+        /// <param name="participant">The participant.</param>
+        void RemoveMigrationParticipant(IGrainMigrationParticipant participant);
+    }
+
+    public interface IGrainMigrationParticipant
```

Comment:
I'm ok with using different tenses. Can you think of a use case for using past/future tense instead of present tense?

To me, past/future tense make the interface seem more like it's for (passive) observers than for (active) participants. I could be in the minority. I also find it slightly confusing: as an implementer, what should I be doing in each of those methods? Should either of them be asynchronous?

Maybe the interface should be renamed to `IGrainLifecycleParticipant` or `IGrainLifecycleFilter` and it can be used as a place to add post-activation hooks, etc, satisfying #4519

As a counter argument (to my argument), you could take precedence from .NET BinaryFormatter which had OnSerializing/OnSerialized/OnDeserializing/OnDeserialized hooks - then again, those aren't the actual serialize/deserialize methods, they are there as some kind of pre/post processing. eg, to grab something from the serialization context after the fields have been deserialized in order to populate some value.

In Orleans, the grain base class has overridable `OnActivateAsync` and `OnDeactivateAsync` methods for the existing lifecycle events, which sets some precedence.

Reply:
My issue with present tense is that it is almost never true. You either fire these callbacks before or after something happens and if it fires in the middle it usually makes a case for it being something more specific. I don't think we need both callback methods but I think its good to communicate WHEN something occurs before or after so that there is no doubt.

Fair point about the existing OnActivateAsync and OnDeactivateAsync. I think provided its documented WHEN it occurs it should be fine.